### PR TITLE
Add mqtt module using paho client

### PR DESCRIPTION
**Pull request by gurbuzali**
#### Which issue this PR fixes
  - fixes #

#### Checklist

- [x] Signed the Hazelcast CLA
- [x] No checkstyle issues (`./gradlew check`)
- [x] No tests failures (`./gradlew test`)
- [ ] Documentation which complies with README template (see `templates/README.template.md`).
- [ ] Update `List of modules` section on the root README with your module.

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/337/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/337/jet-contrib-pr-builder-337.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:47:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:48: Line is longer than 123 characters (found 124). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:51:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:59: Line is longer than 123 characters (found 133). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:80:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:110:25: '500_000' is a magic number. [MagicNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSources.java:45:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSources.java:124:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHooks.java:31:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/package-info.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/338/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/338/jet-contrib-pr-builder-338.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:47:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:48: Line is longer than 123 characters (found 124). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:51:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:59: Line is longer than 123 characters (found 133). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:80:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:110:25: '500_000' is a magic number. [MagicNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSources.java:45:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSources.java:124:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/MqttSerializerHooks.java:31:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/package-info.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/339/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/339/jet-contrib-pr-builder-339.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/HiveMqttSources.java:42:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/HiveMqttSources.java:44:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/HiveMqttSources.java:49:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/HiveMqttSources.java:58: Line is longer than 123 characters (found 124). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/HiveMqttSources.java:74:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSinks.java:42:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSinks.java:43: Line is longer than 123 characters (found 124). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSinks.java:46:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSinks.java:54: Line is longer than 123 characters (found 133). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSinks.java:67:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSources.java:41:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/PahoMqttSources.java:112:9: Redundant 'public' modifier. [RedundantModifier]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:34:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:38:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:42:5: Missing a Javadoc comment. [JavadocMethod]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:54: Missing a Javadoc comment. [JavadocType]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:55:9: Missing a Javadoc comment. [JavadocVariable]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:56:9: Missing a Javadoc comment. [JavadocVariable]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/Subscription.java:57:9: Missing a Javadoc comment. [JavadocVariable]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/paho/SinkCallback.java:43:29: Empty statement. [EmptyStatement]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/paho/SourceCallback.java:40:47: '1024' is a magic number. [MagicNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/impl/paho/SourceCallback.java:41:43: '1024' is a magic number. [MagicNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/jet-contrib-pr-builder/mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/package-info.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
</details>
</pre>

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+     * <p>
+     * By default the client will not attempt to reconnect.
+     * If {@link #connectOptionsFn(SupplierEx)} is set, this parameter is
+     * ignored.
+     * <p>
+     * See {@link MqttConnectOptions#setAutomaticReconnect(boolean)}.
+     */
+    @Nonnull
+    public MqttBaseBuilder<T> autoReconnect() {
+        this.autoReconnect = true;
```
Comment (jbartok) [inline mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java:109]: Automatic reconnection is only one of the reconnect behaviour options offered by `org.eclipse.paho.client.mqttv3.MqttConnectOptions`. Maybe we could cover more and expose our version of the setting as a `com.hazelcast.jet.retry.RetryStrategy`, even if with some limitations?

**Review by jbartok — CHANGES_REQUESTED** [review]

File: mqtt/build.gradle [hunk 2]
```diff
@@ -0,0 +1,8 @@
+dependencies {
+    compile 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.4'
+    testCompile "org.testcontainers:testcontainers:1.15.0-rc2"
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/build.gradle:3]: the tests were failing with a weird docker error and I've found some comments on the internet that this rc release fixes it. It actually fixed the error on my side but I'm not sure if it is related with my environment or not. can you also run the tests on your machine with the stable version ?

File: mqtt/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# MQTT Connector
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/README.md:1]: will do

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+import javax.annotation.Nonnull;
+
+/**
+ * Base class for Mqtt source and sink builder.
+ * See {@link MqttSourceBuilder} and {@link MqttSinkBuilder}.
+ *
+ * @since 4.3
+ */
+abstract class MqttBaseBuilder<T> {
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java]: 👍

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinkBuilder.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    MqttSinkBuilder() {
+    }
+
+    /**
+     * See {@link MqttBaseBuilder#broker(String)}.
+     */
+    @Nonnull
+    @Override
+    public MqttSinkBuilder<T> broker(@Nonnull String broker) {
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinkBuilder.java]: 👍

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinkBuilder.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    @Override
+    public MqttSinkBuilder<T> connectOptionsFn(@Nonnull SupplierEx<MqttConnectOptions> connectOptionsFn) {
+        return (MqttSinkBuilder<T>) super.connectOptionsFn(connectOptionsFn);
+    }
+
+    /**
+     * Set the topic which the sink will publish to. This parameter is
+     * required.
+     * <p>
+     * For example, to publish to the house topic:
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinkBuilder.java:48]: underlying client checks it and I've added a sentence to the javadoc stating that it should not contain wildcards.

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+     * For example:
+     * <pre>{@code
+     * pipeline.writeTo(MqttSinks.publish("topic"));
+     * }</pre>
+     *
+     * @param topic the topic which the sink publishes.
+     * @param <T>   the type of the pipeline item.
+     */
+    @Nonnull
+    public static <T> Sink<T> publish(String topic) {
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:73]: I'm open to suggestions here, currently it is `Source#subscribe` and `Sink#publish`

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java (hunk 1 above)
Comment (gurbuzali) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java]: for a more custom configuration user can always use `connectOptionsFn` and provide their own `MqttConnectOptions` instance. In that case other connection related options such as `autoReconnect`, `keepSession` and `auth` will be ignored.

**Comment by hz-highfive** [comment]
The job `jet-contrib-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/378/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-contrib-pr-builder/378/jet-contrib-pr-builder-378.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

File: mqtt/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbartok) [reply mqtt/build.gradle:3]: `MqttSinkTest.test_retryCount` occasionally fails for me locally.

Have done a few local runs with testcontainers version `1.14.3`, have not seen any errors.

`PahoClientFaultToleranceTest`, shouldn't we do something with it? Delete or fix/finish? Leaving it as `@Ignored` is weird.

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jbartok) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttSinks.java:73]: `MqttSinks.mqtt` and `MqttSources.mqtt` is what would be consistent with other sources and sinks we have. Has it's own weirdness, but hey, at least it's consistent.

File: mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java (hunk 1 above)
Comment (jbartok) [reply mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java]: I still think that we should use `RetryStrategy` in our own facace, the source/sink builders.

File: mqtt/src/test/java/com/hazelcast/jet/contrib/mqtt/MqttSinkTest.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+        p.readFrom(TestSources.items(range(0, itemCount).boxed().collect(toList())))
+                .rebalance()
+                .writeTo(sink);
+
+        jet.newJob(p);
+
+        assertTrueEventually(() -> assertTrue(map.size() > itemCount / 2));
+
+        System.out.println("restart container");
```
Comment (jbartok) [inline mqtt/src/test/java/com/hazelcast/jet/contrib/mqtt/MqttSinkTest.java:94]: Accidental leftover?

**Review by jbartok — APPROVED** [review]

**Pull request merged**